compiler/src/iree/compiler/GlobalOptimization/LiftGenericToTransposeBatchMatmul.cpp

```diff
@@ -0,0 +1,423 @@
+    return *(map.getResultPosition(rewriter.getAffineDimExpr(dimIndex)));
+  };
+  // Permutation from GenericOp lhs shape to BatchMatmulOp lhs shape
+  SmallVector<int64_t> lhsPerm;
+  lhsPerm.push_back(getResultIndex(lhsMap, contractionDims.batch[0]));
+  lhsPerm.push_back(getResultIndex(lhsMap, contractionDims.m[0]));
+  lhsPerm.push_back(getResultIndex(lhsMap, contractionDims.k[0]));
+  // Permutation from GenericOp rhs shape to BatchMatmulOp rhs shape
+  SmallVector<int64_t> rhsPerm;
+  rhsPerm.push_back(getResultIndex(rhsMap, contractionDims.batch[0]));
+  rhsPerm.push_back(getResultIndex(rhsMap, contractionDims.k[0]));
+  rhsPerm.push_back(getResultIndex(rhsMap, contractionDims.n[0]));
+  // Permutation from BatchMatmulOp result shape to GenericOp result shape
+  SmallVector<int64_t> outPerm;
+  outPerm.push_back(getResultIndex(outMap, contractionDims.batch[0]));
+  outPerm.push_back(getResultIndex(outMap, contractionDims.m[0]));
+  outPerm.push_back(getResultIndex(outMap, contractionDims.n[0]));
+  outPerm = invertPermutationVector(outPerm);
+  return liftGenericOp<linalg::BatchMatmulOp>(rewriter, genericOp, lhsPerm,
+                                              rhsPerm, outPerm);
+}
+
+// Converts linalg.conv_2d_input_nhwc_filter_nhwc op to linalg.matmul
+class LiftGenericToTransposeBatchMatmul
+    : public OpRewritePattern<linalg::GenericOp> {
+public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    LLVM_DEBUG(llvm::dbgs() << "LiftGenericToTransposeBatchMatmul on "
+                            << genericOp << "\n\n\n");
+
+    FailureOr<linalg::ContractionDimensions> contractionDims =
+        linalg::inferContractionDims(genericOp);
+    if (failed(contractionDims)) {
+      LLVM_DEBUG(llvm::dbgs() << "failed to infer contraction dims\n\n");
+      return failure();
+    }
+
+    LLVM_DEBUG(llvm::dbgs() << "batch: " << contractionDims->batch << "\n");
+    LLVM_DEBUG(llvm::dbgs() << "m: " << contractionDims->m << "\n");
+    LLVM_DEBUG(llvm::dbgs() << "n: " << contractionDims->n << "\n");
+    LLVM_DEBUG(llvm::dbgs() << "k: " << contractionDims->k << "\n");
+
+    auto lhsType =
+        dyn_cast<RankedTensorType>(genericOp.getOperands()[0].getType());
+    auto rhsType =
+        dyn_cast<RankedTensorType>(genericOp.getOperands()[1].getType());
+    auto outType =
+        dyn_cast<RankedTensorType>(genericOp.getResults()[0].getType());
+    if (!lhsType || !rhsType || !outType) {
+      LLVM_DEBUG(llvm::dbgs() << "Operands do not have RankedTensorType\n\n");
+      return failure();
+    }
+
+    if (!hasMatmulBody(genericOp)) {
+      LLVM_DEBUG(llvm::dbgs() << "genericOp does not have a matmul body\n\n");
+      return failure();
+    }
+
+    if (contractionDims->batch.size() == 1 && contractionDims->m.size() == 0 &&
```

Comment:
Please file and Issue and reference it in a `// TODO(#12345)` comment here about a follow-up addition similar cases for `contractionDims->batch.size() == 0`, creating non-`batch_` ops similar to the `batch_` ops created here.  It's not on our immediate critical path so OK to defer to a TODO, but it'll greatly expand the impact of what you're doing here since in practice there are more matmuls than batch_matmuls outside of the case we're immediately motivated by at the moment.